File: src/app/api/video/providers/veo3.ts
// Veo 3 Provider API Integration
// Server-side logic for Google Veo 3 video generation

import { GoogleGenAI } from "@google/genai";
import sharp from "sharp";
import { s3Client } from "@/utils/s3";
import { PutObjectCommand } from "@aws-sdk/client-s3";
import { getVideoUrl } from "@/utils/urls";
import { fetchImageAsBase64 as fetchImageAsBase64WithMime } from "@/utils/fetchImage";
import { assertAllowedUrl, assertSafePathSegment, encodePathPreservingSlashes } from "@/utils/urlSafety";

const VIDEOS_BUCKET_NAME = process.env.VIDEOS_BUCKET_NAME;

// Veo 3 constraints
const VEO3_VALID_DURATIONS = [4, 6, 8];
const VEO3_VALID_ASPECT_RATIOS = ["16:9", "9:16"];

export interface Veo3SubmitRequest {
  prompt: string;
  imageBase64?: string;
  duration: number;
  aspectRatio: "16:9" | "9:16";
}

export interface Veo3SubmitResult {
  jobId: string;
  status: "pending";
}

export interface Veo3StatusResult {
  jobId: string;
  status: "pending" | "running" | "completed" | "failed";
  videoUrl?: string;
  s3FileName?: string;
  error?: string;
}

/**
 * Check if string is a URL
 */
function isUrl(str: string): boolean {
  return str.startsWith("http://") || str.startsWith("https://");
}

/**
 * Resize image to exact Veo 3 dimensions
 */
async function resizeImageToVeo3Dimensions(
  imageInput: string,
  width: number,
  height: number
): Promise<{ base64: string; mimeType: string }> {
  // Handle both S3 URLs and base64 strings
  let base64: string;
  if (isUrl(imageInput)) {
    ({ base64 } = await fetchImageAsBase64WithMime(imageInput));
  } else {
    base64 = imageInput;
  }

  const buffer = Buffer.from(base64, "base64");
  const resized = await sharp(buffer)
    .resize(width, height, { fit: "cover" })
    .jpeg({ quality: 90 })
    .toBuffer();
  return {
    base64: resized.toString("base64"),
    mimeType: "image/jpeg",
  };
}

/**
 * Map duration to nearest valid Veo 3 duration
 */
export function mapToVeo3Duration(duration: number): number {
  return VEO3_VALID_DURATIONS.reduce((prev, curr) =>
    Math.abs(curr - duration) < Math.abs(prev - duration) ? curr : prev
  );
}

/**
 * Validate Veo 3 request
 */
export function validateVeo3Request(request: Veo3SubmitRequest): string | null {
  if (!request.prompt || typeof request.prompt !== "string") {
    return "Prompt is required and must be a string";
  }

  if (
    !request.aspectRatio ||
    !VEO3_VALID_ASPECT_RATIOS.includes(request.aspectRatio)
  ) {
    return "Aspect ratio must be '16:9' or '9:16'";
  }

  return null;
}

/**
 * Submit a video generation job to Veo 3
 */
export async function submitToVeo3(
  request: Veo3SubmitRequest,
  customApiKey?: string
): Promise<Veo3SubmitResult> {
  const apiKey = customApiKey || process.env.GEMINI_API_KEY;
  if (!apiKey) {
    throw new Error("GEMINI_API_KEY is not configured");
  }

  const { prompt, imageBase64, duration, aspectRatio } = request;

  // Validate
  const validationError = validateVeo3Request(request);
  if (validationError) {
    throw new Error(validationError);
  }

  // Map to valid Veo 3 duration
  const veo3Duration = mapToVeo3Duration(duration);

  const ai = new GoogleGenAI({ apiKey });

  // Build request config
  const config: {
    aspectRatio: string;
    durationSeconds: number;
    personGeneration: string;
  } = {
    aspectRatio,
    durationSeconds: veo3Duration,
    personGeneration: "allow_adult",
  };

  // Build request with or without image
  const requestBody: Record<string, unknown> = {
    model: "veo-3.1-generate-preview",
    prompt,
    config,
  };

  // Add image reference if provided (image-to-video)
  if (imageBase64) {
    const targetWidth = aspectRatio === "16:9" ? 1280 : 720;
    const targetHeight = aspectRatio === "16:9" ? 720 : 1280;

    const { base64, mimeType } = await resizeImageToVeo3Dimensions(
      imageBase64,
      targetWidth,
      targetHeight
    );

    requestBody.image = {
      imageBytes: base64,
      mimeType,
    };
  }


  // Submit the video generation job
  // Cast to unknown first to avoid type incompatibility
  const models = ai.models as unknown as { generateVideos: (body: unknown) => Promise<{ name?: string; operationName?: string; id?: string; operationId?: string }> };
  const operation = await models.generateVideos(requestBody);


  // The operation.name is the job ID we'll use to poll status
  // Try different property names that the SDK might use
  const jobId = operation.name || operation.operationName || operation.id || operation.operationId;

  if (!jobId) {
    console.error("[veo3] Operation object keys:", Object.keys(operation || {}));
    throw new Error("Failed to get operation ID from Veo 3 response");
  }


  return {
    jobId,
    status: "pending",
  };
}

/**
 * Check the status of a Veo 3 video generation job
 * If completed, downloads from Google and uploads to S3
 */
export async function checkVeo3Status(
  jobId: string,
  customApiKey?: string
): Promise<Veo3StatusResult> {
  const safeJobId = assertSafePathSegment(jobId, { allowSlash: true });
  const apiKey = customApiKey || process.env.GEMINI_API_KEY;
  if (!apiKey) {
    throw new Error("GEMINI_API_KEY is not configured");
  }

  const encodedJobId = encodePathPreservingSlashes(safeJobId);

  // Use REST API directly to poll operation status
  // The SDK methods don't work reliably for operation polling
  const response = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/${encodedJobId}?key=${apiKey}`,
    {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    }
  );

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    console.error("[veo3] REST API error:", response.status, errorData);
    throw new Error(
      errorData.error?.message || `Failed to check operation status: ${response.status}`
    );
  }

  const operation = await response.json();

  // Determine status
  let status: "pending" | "running" | "completed" | "failed";

  if (operation.done) {
    if (operation.error) {
      status = "failed";
    } else {
      status = "completed";
    }
  } else {
    // Operation is still in progress
    status = "running";
  }


  const result: Veo3StatusResult = {
    jobId,
    status,
  };

  // If completed, download from Google and upload to S3
  const generatedSamples = operation.response?.generateVideoResponse?.generatedSamples;
  if (status === "completed" && generatedSamples?.length > 0) {
    try {
      const videoUri = generatedSamples[0].video?.uri;

      if (!videoUri) {
        throw new Error("No video URI in response");
      }

      const safeVideoUrl = assertAllowedUrl(videoUri, {
        allowedHostSuffixes: [".googleapis.com"],
        allowedHostnames: new Set(),
      });

      // Download the video directly from the URI
      // Add API key if not already in the URL
      const safeVideoUri = safeVideoUrl.toString();
      const downloadUrl = safeVideoUri.includes("key=")
        ? safeVideoUri
        : `${safeVideoUri}&key=${apiKey}`;

      const videoResponse = await fetch(downloadUrl);

      if (!videoResponse.ok) {
        throw new Error(`Failed to download video: ${videoResponse.status} ${videoResponse.statusText}`);
      }

      const videoArrayBuffer = await videoResponse.arrayBuffer();
      const videoBuffer = Buffer.from(videoArrayBuffer);


      // Generate unique filename with jobId for traceability
      // Replace slashes in jobId since it contains path-like structure
      const sanitizedJobId = jobId.replace(/\//g, "_");
      const s3FileName = `veo3_${Date.now()}_${sanitizedJobId}.mp4`;

      // Upload to S3
      const uploadParams = {
        Bucket: VIDEOS_BUCKET_NAME,
        Key: s3FileName,
        Body: videoBuffer,
        ContentType: "video/mp4",
      };

      await s3Client.send(new PutObjectCommand(uploadParams));

      // Return CloudFront URL for permanent access
      result.videoUrl = getVideoUrl(s3FileName);
      result.s3FileName = s3FileName;
    } catch (uploadError) {
      console.error("[veo3] Error uploading to S3:", uploadError);
      throw uploadError;
    }
  }

  // If failed, include the error message
  if (status === "failed") {
    result.error =
      operation.error?.message || "Video generation failed";
    console.error("[veo3] Video generation failed:", result.error);
  }

  return result;
}

File: mithril-backend/app/providers/veo3.py
"""Veo 3 (Google) video generation provider using google-genai SDK."""

import asyncio
import logging

import httpx
from google import genai

from app.models.provider import (
    PollingConfig,
    ProviderConstraints,
    VideoSubmitRequest,
    VideoSubmitResult,
    VideoStatusResult,
)
from app.providers.base import VideoProvider
from app.services.image_processor import prepare_image_for_provider

logger = logging.getLogger(__name__)

# Limit concurrent Veo3 API calls per worker to avoid rate-limit errors
_VEO3_SEMAPHORE = asyncio.Semaphore(2)

# Veo 3 API base URL (for status polling)
VEO3_API_BASE = "https://generativelanguage.googleapis.com/v1beta"

# Veo 3 constraints
VEO3_CONSTRAINTS = ProviderConstraints(
    durations=[4, 6, 8],
    sizes={
        "16:9": "1280x720",
        "9:16": "720x1280",
    },
    polling=PollingConfig(
        interval_ms=10000,
        max_attempts=90,
    ),
    supports_image_to_video=True,
)


class Veo3Provider(VideoProvider):
    """Google Veo 3 video generation provider."""

    @property
    def id(self) -> str:
        return "veo3"

    @property
    def name(self) -> str:
        return "Veo 3"

    @property
    def description(self) -> str:
        return "Google's Veo 3 video generation model"

    @property
    def model_name(self) -> str:
        return "veo-3.1-fast-generate-preview"

    def get_constraints(self) -> ProviderConstraints:
        return VEO3_CONSTRAINTS

    async def submit_job(
        self,
        request: VideoSubmitRequest,
        api_key: str,
    ) -> VideoSubmitResult:
        """Submit video generation job to Veo 3 using SDK."""
        # Validate request
        validation_error = self.validate_request(request)
        if validation_error:
            raise ValueError(validation_error)

        # Map duration to valid Veo 3 duration
        duration = self.map_duration(request.duration)

        logger.info(f"[veo3] Submitting video generation job...")
        logger.info(f"[veo3] Model: {self.model_name}")
        logger.info(f"[veo3] Duration: {duration} seconds")
        logger.info(f"[veo3] Aspect ratio: {request.aspect_ratio}")
        logger.info(f"[veo3] Has start image: {bool(request.image_base64 or request.image_url)}")
        logger.info(f"[veo3] Has end image: {bool(request.image_end_url)}")

        # Initialize the SDK client
        client = genai.Client(api_key=api_key)

        # Build config
        config = {
            "aspect_ratio": request.aspect_ratio,
            "duration_seconds": duration,
            "person_generation": "allow_adult",
        }

        # Prepare start image if provided
        image_data = None
        if request.image_base64 or request.image_url:
            resized_base64 = await prepare_image_for_provider(
                request.image_base64,
                request.image_url,
                request.aspect_ratio,
            )
            if resized_base64:
                image_data = {
                    "image_bytes": resized_base64,
                    "mime_type": "image/jpeg",
                }

        # Prepare end image if provided
        image_end_data = None
        if request.image_end_url:
            resized_end_base64 = await prepare_image_for_provider(
                None,
                request.image_end_url,
                request.aspect_ratio,
            )
            if resized_end_base64:
                image_end_data = {
                    "image_bytes": resized_end_base64,
                    "mime_type": "image/jpeg",
                }

        logger.info(f"[veo3] image_data present: {image_data is not None}")
        logger.info(f"[veo3] image_end_data present: {image_end_data is not None}")
        logger.info(f"[veo3] config before submit: { {k: v for k, v in config.items() if k != 'last_frame'} }")

        try:
            # Use the SDK to generate video (run sync SDK in thread pool)
            def _submit():
                kwargs = {
                    "model": self.model_name,
                    "prompt": request.prompt,
                    "config": config,
                }
                if image_data:
                    kwargs["image"] = image_data
                    # lastFrame requires image (start frame) — camelCase matches REST API
                    if image_end_data:
                        config["lastFrame"] = image_end_data
                logger.info(f"[veo3] generate_videos kwargs keys: {list(kwargs.keys())}")
                logger.info(f"[veo3] config keys: {list(config.keys())}")
                return client.models.generate_videos(**kwargs)

            async with _VEO3_SEMAPHORE:
                operation = await asyncio.to_thread(_submit)

            logger.info(f"[veo3] Operation response: {operation}")

            # Get the operation name (job ID)
            job_id = getattr(operation, 'name', None) or getattr(operation, 'operation_name', None)

            if not job_id:
                # Try to get from dict if it's a dict-like object
                if hasattr(operation, '__getitem__'):
                    job_id = operation.get('name') or operation.get('operationName')

            if not job_id:
                logger.error(f"[veo3] Operation object: {operation}")
                raise Exception("Failed to get operation ID from Veo 3 response")

            logger.info(f"[veo3] Job submitted successfully. Operation name: {job_id}")
            return VideoSubmitResult(job_id=job_id, status="pending")

        except Exception as e:
            error_msg = str(e)
            logger.error(f"[veo3] Exception type: {type(e).__name__}")
            logger.error(f"[veo3] Full error: {e}")
            if hasattr(e, '__dict__'):
                logger.error(f"[veo3] Error attrs: {e.__dict__}")
            if "rate limit" in error_msg.lower() or "429" in error_msg:
                raise Exception("Rate limit exceeded. Please try again later.")
            if "quota" in error_msg.lower() or "402" in error_msg:
                raise Exception("Quota exceeded. Please check your API usage.")
            logger.error(f"[veo3] Error submitting job: {e}")
            raise

    async def check_status(
        self,
        job_id: str,
        api_key: str,
    ) -> VideoStatusResult:
        """Check the status of a Veo 3 video generation job."""
        logger.info(f"[veo3] Checking status for operation: {job_id}")

        # Use REST API to poll operation status (SDK polling is unreliable)
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.get(
                f"{VEO3_API_BASE}/{job_id}",
                params={"key": api_key},
                headers={"Content-Type": "application/json"},
            )

            if response.status_code == 404:
                raise Exception("Job not found. It may have expired.")

            response_data = response.json()

            if not response.is_success:
                error_msg = (
                    response_data.get("error", {}).get("message")
                    or f"Failed to check operation status: {response.status_code}"
                )
                logger.error(f"[veo3] REST API error: {response.status_code} {response_data}")
                raise Exception(error_msg)

            logger.debug(f"[veo3] Operation response: {response_data}")

            # Determine status
            if response_data.get("done"):
                if response_data.get("error"):
                    status = "failed"
                else:
                    status = "completed"
            else:
                status = "running"

            logger.info(f"[veo3] Operation status: {status}, done: {response_data.get('done')}")

            result = VideoStatusResult(job_id=job_id, status=status)

            # If failed, include error message
            if status == "failed":
                result.error = (
                    response_data.get("error", {}).get("message")
                    or "Video generation failed"
                )
                logger.error(f"[veo3] Video generation failed: {result.error}")

            return result

    async def download_video(
        self,
        job_id: str,
        api_key: str,
    ) -> bytes:
        """Download completed video from Veo 3."""
        # First, get the operation to find the video URI
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.get(
                f"{VEO3_API_BASE}/{job_id}",
                params={"key": api_key},
                headers={"Content-Type": "application/json"},
            )

            if not response.is_success:
                raise Exception(f"Failed to get operation: {response.status_code}")

            operation = response.json()

            # Extract video URI from response
            generated_samples = (
                operation.get("response", {})
                .get("generateVideoResponse", {})
                .get("generatedSamples", [])
            )

            if not generated_samples:
                raise Exception("No generated samples in response")

            video_uri = generated_samples[0].get("video", {}).get("uri")
            if not video_uri:
                raise Exception("No video URI in response")

            logger.info(f"[veo3] Downloading video from: {video_uri}")

            # Download the video
            # Add API key if not already in URL
            download_url = (
                video_uri if "key=" in video_uri else f"{video_uri}&key={api_key}"
            )

        async with httpx.AsyncClient(timeout=120.0, follow_redirects=True) as client:
            video_response = await client.get(download_url)

            if not video_response.is_success:
                raise Exception(
                    f"Failed to download video: {video_response.status_code}"
                )

            logger.info(f"[veo3] Video downloaded, size: {len(video_response.content)} bytes")
            return video_response.content


# Create singleton instance
veo3_provider = Veo3Provider()